### ISSUE TITLE
TODO list of supporting PyTorch

### ISSUE THREAD

**Issue by QiJune** [sql-machine-learning/elasticdl#2149]
### 功能支持

-  forward/optimizer/loss/feed function in model definition
- data loader module
- implement a training loop: (get task --> read minibatch --> pull parameters -->  forward --> backward --> push gradients)

### 重构工作

PyTorch 和 TensorFlow 在worker端的整体逻辑基本一致，需要考虑如何做模块划分，复用代码。
比如 worker 与 parameter server 之间通信的接口可以独立出来，是能复用的；再比如 worker 从 master 获取 task 的逻辑也是可以复用的。

## master开发

- evaluation function

## 验证与总结

- 在minikube上验证
- 在google cloud上验证
- 撰写稿件

**Comment by skydoorkai** [comment]
1. Load/Save checkpoint (in PS and worker)
1. Embedding layer support
1. nn.Module model zoo

**Comment by QiJune** [comment]
Step 1：  重构与代码清理工作

worker 与 PS之间的通信：

- get parameter
- report gradient

worker 与 master 之间的通信：

- get task
- report evaluation

这些接口都是可以复用的，我们先剥离出一个 client出来，用于封装这些rpc通信接口，供TensorFlow 和 PyTorch来共同调用。

拆分之后的代码结构应该是这个样子：

```
worker
    worker_pytorch.py
    worker_tf.py
    ps_client.py
    master_client.py
```

**Comment by Yancey0623** [comment]
# Work Plan of ElasticDL Supporting PyTorch

We would like to publish documentation to introduce ElasticDL supporting PyTorch and contribute to some media at the end of this work.

The current codebase only supported TensorFlow, to make it easy to support other frameworks e.g. PyTorch and XGBoost in the future, we should do much more refactoring work to make the system design clearer, so the refactoring is also within the scope of this work.

## Weekly Plan

The project duration: 5 weeks.

Week1:
- Code refactor (@QiJune @Yancey1989 )
    - decouple worker, master, and pserver APIs.
    - design: supporting different framework e.g. TensorFlow, PyTorch, and XGBoost.
- Submit the local PyTorch training job with ElasticDL command-line (@Kelang-Tian )

Week2:
- PyTorch worker design(@Kelang-Tian)
- PyTorch worker supports training with custom PyTorch model (@Kelang-Tian)
-  Code refactor (@QiJune , @Yancey1989 )
    - Support different distributed strategy: parameter server and collective


Week3:
- PyTorch worker develop(@QiJune @Kelang-Tian @Yancey1989 )
    - Train
    - Evalution
    - Prediction

Week4:
- PyTorch worker support collective mode(@QiJune @Kelang-Tian  @Yancey1989 )

Week5:
- Experiment on Google Cloud and write down the documentation. @Kelang-Tian
